### DISCUSSION TITLE
Multi transport approach

### DISCUSSION THREAD

**Discussion by sfavazza** [capnproto/capnproto#1900] [Q&A]
I am quite new in Cap'n Proto, and besides managing to bring up a simple Ez-based client/server approach I was wondering:
- can a server serve over multiple transports at the same time? e.g. :
  - a Unix socket
  - another socket type (e.g. to a PCIe interface)
- if yes, is there any chance of race condition when two clients (connected from different sockets) request the same service?

What would be the best starting point to implement an "exotic" transport? (I was scanning through the ez-rpc.c++ code and from my understanding I should focus on `kj::AsyncIoContext` and `kj::setupAsyncIo`).

**Comment by kentonv** [comment]
Yes, a single server can wait on many sockets. Services are designed to have many clients, using event loop concurrency.

I recommend steering away from EZ RPC as it turns out to hide too much. The right approach is to use `setupAsyncIo()` and `capnp::TwoPartyServer`.

KJ only knows how to create TCP and Unix sockets. But, using `kj::LowLevelAsyncIoProvider` (which is provided as part of `AsyncIoContext`), you can wrap arbitrary socket file descriptors. As long as they behave like any other stream-oriented socket, it should just work. So you can create your PCIe socket manually, then wrap it, assuming PCIe sockets are stream-oriented (I have no idea).
